src/lib/accessibility/FocusZone/FocusTrapZone.tsx

```diff
@@ -0,0 +1,322 @@
+      ev.preventDefault()
+      ev.stopPropagation()
+    } else if (!ev.shiftKey && _lastTabbableChild === ev.target) {
+      focusAsync(_firstTabbableChild)
+      ev.preventDefault()
+      ev.stopPropagation()
+    }
+  }
+
+  private _forceFocusInTrap = (ev: FocusEvent): void => {
+    if (
+      FocusTrapZone._focusStack.length &&
+      this === FocusTrapZone._focusStack[FocusTrapZone._focusStack.length - 1]
+    ) {
+      const focusedElement = document.activeElement as HTMLElement
+
+      if (!this._root.current.contains(focusedElement)) {
+        this.focus()
+        ev.preventDefault()
+        ev.stopPropagation()
+      }
+    }
+  }
+
+  private _forceClickInTrap = (ev: MouseEvent): void => {
+    if (
+      FocusTrapZone._focusStack.length &&
+      this === FocusTrapZone._focusStack[FocusTrapZone._focusStack.length - 1]
+    ) {
+      const clickedElement = ev.target as HTMLElement
+
+      if (clickedElement && !this._root.current.contains(clickedElement)) {
+        this.focus()
+        ev.preventDefault()
+        ev.stopPropagation()
+      }
+    }
+  }
+
+  private _subscribeToEvents = () => {
+    const { forceFocusInsideTrap, isClickableOutsideFocusTrap } = this.props
+    if (forceFocusInsideTrap) {
+      eventStack.sub('focus', this._forceFocusInTrap, {
+        target: this.windowElement,
+        useCapture: true,
+      })
+    }
+
+    if (!isClickableOutsideFocusTrap) {
+      eventStack.sub('click', this._forceClickInTrap, {
+        target: this.windowElement,
+        useCapture: true,
+      })
+    }
+  }
+
+  private _unsubscribeFromEvents = () => {
+    const { forceFocusInsideTrap, isClickableOutsideFocusTrap } = this.props
+    if (forceFocusInsideTrap) {
+      eventStack.unsub('focus', this._forceFocusInTrap, {
+        target: this.windowElement,
+        useCapture: true,
+      })
+    }
+
+    if (!isClickableOutsideFocusTrap) {
+      eventStack.unsub('click', this._forceClickInTrap, {
+        target: this.windowElement,
+        useCapture: true,
+      })
+    }
+  }
+
+  private _getPreviouslyFocusedElementOutsideTrapZone = () => {
+    const { elementToFocusOnDismiss } = this.props
+    let previouslyFocusedElement = this._previouslyFocusedElementOutsideTrapZone
+
+    if (elementToFocusOnDismiss && previouslyFocusedElement !== elementToFocusOnDismiss) {
+      previouslyFocusedElement = elementToFocusOnDismiss
+    } else if (!previouslyFocusedElement) {
+      previouslyFocusedElement = document.activeElement as HTMLElement
+    }
+
+    return previouslyFocusedElement
+  }
+
+  private _hideContentFromAccessibilityTree = () => {
+    const elements = (document.body && document.body.children) || []
+
+    if (elements.length && !document.body.contains(this._root.current)) {
+      // In case popup render options will change
+      throw new Error(
+        'Body does not contain trap zone element as expected. If it is done intentionally, please, make sure to update FocusTrapZone.',
```

Comment:
sorry - could you, please, elaborate a bit further on what could be the case for it? Just describe the one where you see this could happen. Thanks!

Reply:
This one we agreed with @miroslavstastny. So in case, someone will use Portal outside of the body, we have to notify that FocusTrapZone would not work in that case.

Reply:
ok - but in that case shouldn't it be provided as a warning? The thing is that current behavior will result in failed rendering for React, and there won't be **_any_** tree rendered.

Reply:
agree with you, I will change it to a warning, to not interrupt rendering of React